openedx_external_enrollments/external_enrollments.py

```diff
@@ -0,0 +1,424 @@
+        json_data = self._get_enrollment_data(data, course_settings)
+        LOG.info('calling enrollment for [%s] with data: %s', self.__str__(), json_data)
+        LOG.info('calling enrollment for [%s] with url: %s', self.__str__(), url)
+        LOG.info('calling enrollment for [%s] with course settings: %s', self.__str__(), course_settings)
+        log_details = {
+            "request_payload": json_data,
+            "url": url,
+            "course_advanced_settings": course_settings,
+        }
+
+        try:
+            response = self._execute_post(
+                url=url,
+                headers=self._get_enrollment_headers(),
+                json_data=json_data,
+            )
+        except Exception as error:
+            LOG.error("Failed to complete enrollment. Reason: " + str(error))
+            log_details["response"] = {"error": "Failed to complete enrollment. Reason: " + str(error)}
+            EnrollmentRequestLog.objects.create(request_type=str(self), details=log_details)
+            return str(error), status.HTTP_400_BAD_REQUEST
+        else:
+            LOG.info('External enrollment response for [%s] -- %s', self.__str__(), response.json())
+            log_details["response"] = response.json()
+            EnrollmentRequestLog.objects.create(request_type=str(self), details=log_details)
+            return response.json(), status.HTTP_200_OK
+
+    def _get_enrollment_data(course_settings):
+        """Unimplemented method necessary to execute _post_enrollment."""
+        raise NotImplementedError
+
+    def _get_enrollment_headers(course_settings):
+        """Unimplemented method necessary to execute _post_enrollment."""
+        raise NotImplementedError
+
+    def _get_enrollment_url(course_settings):
+        """Unimplemented method necessary to execute _post_enrollment."""
+        raise NotImplementedError
+
+
+class EdxEnterpriseExternalEnrollment(BaseExternalEnrollment):
+    """
+    """
+
+    def __str__(self):
+        return "edX"
+
+    def _get_enrollment_headers(self):
+        """
+        """
+        try:
+            data = OrderedDict(
+                grant_type="client_credentials",
+                client_id=settings.EDX_ENTERPRISE_API_CLIENT_ID,
+                client_secret=settings.EDX_ENTERPRISE_API_CLIENT_SECRET,
+                token_type="jwt",
+            )
+            response = self._execute_post(
+                settings.EDX_ENTERPRISE_API_TOKEN_URL,
+                data,
+            )
+        except Exception as error:
+            LOG.error("Failed to get token: " + str(error))
+        else:
+            if response.ok:
+                headers = {
+                    "Accept": "application/json",
+                    "Content-Type": "application/json",
+                    "Authorization": "{} {}".format(
+                        response.json()["token_type"],
+                        response.json()["access_token"]
```

Comment:
Safe indexing. Trailing comma. @andrey-canon

Reply:
out of the scope of this pr